hail/src/main/scala/is/hail/expr/LowerArrayToStream.scala

```diff
@@ -0,0 +1,65 @@
+package is.hail.expr.ir
+
+import is.hail.expr.types.virtual.{TArray, TContainer, TIterable, TStream}
+
+object LowerArrayToStream {
+  private def nonStreamIterable(node: IR) = node.typ.isInstanceOf[TIterable] && !node.typ.isInstanceOf[TStream]
+
+  private def boundary(node: IR): IR = {
+    val streamified = streamify(node)
+
+    val wrapped = if (streamified.typ.isInstanceOf[TStream] && node.typ.isInstanceOf[TArray])
+      ToArray(streamified)
+    else
+      streamified
+
+    assert(wrapped.typ == node.typ)
+    wrapped
+  }
+
+  private def toStream(node: IR): IR = {
+    val streamified = streamify(node)
+
+    if (nonStreamIterable(streamified))
+      ToStream(streamified)
+    else
+      streamified
+  }
+
+  private def streamify(node: IR): IR = {
+    node match {
+      case ArraySort(a, l, r, comp) => ArraySort(toStream(a), l, r, boundary(comp))
+      case ArrayFold(a, zero, zn, an, body) => ArrayFold(toStream(a), boundary(zero), zn, an, boundary(body))
+      case ArrayFor(a, n, b) => ArrayFor(toStream(a), n, boundary(b))
+      case ArrayFold2(a, accum, valueName, seq, result) => ArrayFold2(
+        toStream(a),
+        accum.map { case (name, value) => (name, boundary(value)) },
+        valueName, seq.map(boundary), boundary(result))
+      case RunAggScan(a, name, init, seq, res, sig) => RunAggScan(toStream(a), name, boundary(init), boundary(seq), boundary(res), sig)
+      case MakeArray(args, t) => MakeStream(args.map(boundary), TStream(t.elementType, t.required))
+      case ArrayRange(start, stop, step) => StreamRange(boundary(start), boundary(stop), boundary(step))
+      case ArrayZip(childIRs, names, body, behavior) => ArrayZip(childIRs.map(toStream), names, boundary(body), behavior)
+      case ArrayMap(a, n, b) => ArrayMap(toStream(a), n, boundary(b))
+      case ArrayFilter(a, n, b) => ArrayFilter(toStream(a), n, boundary(b))
+      case ArrayFlatMap(a, n, b) => ArrayFlatMap(toStream(a), n, toStream(b))
+      case ArrayScan(a, zero, zn, an, body) => ArrayScan(toStream(a), boundary(zero), zn, an, boundary(body))
+      case ArrayLeftJoinDistinct(l, r, ln, rn, keyf, joinf) =>
+        ArrayLeftJoinDistinct(toStream(l), toStream(r), ln, rn, boundary(keyf), boundary(joinf))
+      case x: ApplyIR => streamify(x.explicitNode)
```

Comment:
could you help me understand why this isn't boundary?

Reply:
oh, it should be

Reply:
I think this case shouldn't be here at all

Reply:
yeah exactly, the child will handle

Reply:
er, streamify is right

Reply:
totally

Reply:
(though this should not be here, it should be in another lowering pass)